File: backend/app/routers/journal.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from datetime import date, datetime, timedelta
from pydantic import BaseModel
from app.db import get_db
from app.models.journal import Journal
from app.models.annotation import Annotation

router = APIRouter(prefix="/api/journal", tags=["journal"])


class JournalUpsert(BaseModel):
    trade_date: str
    operations: str = ""
    market_obs: str = ""
    plan: str = ""
    mood: str = "neutral"


def _serialize(j: Journal) -> dict:
    return {
        "id": j.id,
        "trade_date": j.trade_date.isoformat() if j.trade_date else None,
        "operations": j.operations,
        "market_obs": j.market_obs,
        "plan": j.plan,
        "mood": j.mood,
        "created_at": j.created_at.isoformat() if j.created_at else None,
        "updated_at": j.updated_at.isoformat() if j.updated_at else None,
    }


@router.get("")
async def get_journal(date: str = Query(..., description="YYYY-MM-DD"), db: AsyncSession = Depends(get_db)):
    """Get journal entry for a specific date."""
    try:
        d = datetime.strptime(date, "%Y-%m-%d").date()
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid date format, use YYYY-MM-DD")
    j = await db.scalar(select(Journal).where(Journal.trade_date == d))
    if not j:
        return {"trade_date": date, "operations": "", "market_obs": "", "plan": "", "mood": "neutral"}
    return _serialize(j)


@router.put("")
async def upsert_journal(data: JournalUpsert, db: AsyncSession = Depends(get_db)):
    """Create or update journal entry (upsert by trade_date)."""
    try:
        d = datetime.strptime(data.trade_date, "%Y-%m-%d").date()
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid date format, use YYYY-MM-DD")

    j = await db.scalar(select(Journal).where(Journal.trade_date == d))
    if j:
        j.operations = data.operations
        j.market_obs = data.market_obs
        j.plan = data.plan
        j.mood = data.mood
        j.updated_at = datetime.utcnow()
    else:
        j = Journal(
            trade_date=d,
            operations=data.operations,
            market_obs=data.market_obs,
            plan=data.plan,
            mood=data.mood,
        )
        db.add(j)
    await db.commit()
    await db.refresh(j)
    return _serialize(j)


@router.get("/recent")
async def recent_journals(days: int = Query(30, ge=1, le=365), db: AsyncSession = Depends(get_db)):
    """Get recent journal entries (summary list)."""
    cutoff = date.today() - timedelta(days=days)
    q = (
        select(Journal)
        .where(Journal.trade_date >= cutoff)
        .order_by(Journal.trade_date.desc())
    )
    result = await db.execute(q)
    journals = result.scalars().all()
    return [
        {
            "trade_date": j.trade_date.isoformat(),
            "mood": j.mood,
            "summary": (j.operations or "")[:80],
        }
        for j in journals
    ]


@router.get("/{date}/annotations")
async def date_annotations(date: str, db: AsyncSession = Depends(get_db)):
    """Get all annotations across stocks for a given date."""
    q = (
        select(Annotation)
        .where(Annotation.trade_date == date)
        .order_by(Annotation.stock_code)
    )
    result = await db.execute(q)
    annotations = result.scalars().all()
    return [
        {
            "id": a.id,
            "stock_code": a.stock_code,
            "trade_date": a.trade_date,
            "type": a.type,
            "content": a.content,
            "position": a.position,
        }
        for a in annotations
    ]

File: frontend/src/stores/journal.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { journalApi } from '@/api'

export interface JournalEntry {
  id?: number
  trade_date: string
  operations: string
  market_obs: string
  plan: string
  mood: string
  created_at?: string
  updated_at?: string
}

export interface JournalSummary {
  trade_date: string
  mood: string
  summary: string
}

export interface DateAnnotation {
  id: string
  stock_code: string
  trade_date: string
  type: string
  content: string
  position: string
}

export const useJournalStore = defineStore('journal', () => {
  const entry = ref<JournalEntry>({
    trade_date: new Date().toISOString().slice(0, 10),
    operations: '',
    market_obs: '',
    plan: '',
    mood: 'neutral',
  })
  const recent = ref<JournalSummary[]>([])
  const annotations = ref<DateAnnotation[]>([])
  const loading = ref(false)
  const saving = ref(false)

  function todayStr(): string {
    return new Date().toISOString().slice(0, 10)
  }

  async function load(date: string) {
    loading.value = true
    try {
      const data = await journalApi.get(date)
      entry.value = { ...data }
      // Also load annotations for this date
      const annData = await journalApi.annotations(date)
      annotations.value = annData || []
    } finally {
      loading.value = false
    }
  }

  async function save() {
    saving.value = true
    try {
      const data = await journalApi.upsert({
        trade_date: entry.value.trade_date,
        operations: entry.value.operations,
        market_obs: entry.value.market_obs,
        plan: entry.value.plan,
        mood: entry.value.mood,
      })
      entry.value = { ...data }
      // Refresh recent list
      await loadRecent()
      return data
    } finally {
      saving.value = false
    }
  }

  async function loadRecent(days = 30) {
    const data = await journalApi.recent(days)
    recent.value = data || []
  }

  function setDate(date: string) {
    entry.value.trade_date = date
  }

  return {
    entry,
    recent,
    annotations,
    loading,
    saving,
    todayStr,
    load,
    save,
    loadRecent,
    setDate,
  }
})
